### PR TITLE
Add new getLockState() debugging endpoint

### PR DESCRIPTION
**Pull request by dxiao**
Allows for easier, more targetted debugging than would be
possible with previous logCurrentState() endpoint

### CODE REVIEW

File: lock-impl/src/main/java/com/palantir/lock/impl/LockServiceImpl.java [hunk 1]
```diff
@@ -1128,6 +1133,73 @@ public LockServerOptions getLockServerOptions() {
...
+                .build()));
+        requests.forEach((request, client) -> lockState.addRequesters(ImmutableLockRequester.builder()
+                .client(client)
+                .lockGroupBehavior(request.getLockGroupBehavior())
+                .blockingMode(request.getBlockingMode())
+                .blockingDuration(Optional.ofNullable(request.getBlockingDuration()))
+                .versionId(Optional.ofNullable(request.getVersionId()))
+                .requestingThread(request.getCreatingThreadName())
+                .build()));
+        return lockState.build();
```
[the remark this answers is not in the document]
Comment (jeremyk-91) [reply lock-impl/src/main/java/com/palantir/lock/impl/LockServiceImpl.java:1200]: I think factories for creating a LockHolder/LockRequester from the request and/or lock object would be nice, though requests.forEach(...) in the body here seems fine to me

**Review by jeremyk-91 — COMMENTED** [review]
Broadly makes sense, though I'm pretty sure RLSA needs to be different. I think @sudiksha27 had a number of good suggestions.

File: lock-api/src/main/java/com/palantir/lock/client/RemoteLockServiceAdapter.java [hunk 2]
```diff
@@ -157,4 +159,9 @@ public long currentTimeMillis() {
     public void logCurrentState() {
         namespaceAgnosticLockRpcClient.logCurrentState();
     }
+
+    @Override
+    public LockState getLockState(LockDescriptor lock) {
+        return null;
```
[the remark this answers is not in the document]
Comment (dxiao) [reply lock-api/src/main/java/com/palantir/lock/client/RemoteLockServiceAdapter.java:165]: oops!

File: lock-impl/src/main/java/com/palantir/lock/impl/LockServiceImpl.java [hunk 3]
```diff
@@ -1128,6 +1133,73 @@ public LockServerOptions getLockServerOptions() {
...
+        List<LockClient> lockHolders;
+        if (readHolders.isEmpty()) {
+            if (writeHolders == null) {
+                lockHolders = ImmutableList.of();
+            } else {
+                lockHolders = ImmutableList.of(writeHolders);
+            }
+        } else {
+            lockHolders = readHolders;
+        }
```
[the remark this answers is not in the document]
Comment (dxiao) [reply lock-impl/src/main/java/com/palantir/lock/impl/LockServiceImpl.java:1166]: Seems like a lot of fanciness for not a lot of benefit, but either way's fine by me!

File: lock-impl/src/test/java/com/palantir/lock/LockServiceTest.java [hunk 4]
```diff
@@ -520,6 +523,51 @@ public void testLogCurrentState() throws Exception {
...
+            return null;
+        });
+
+        /* Now make the logCurrentState() request; with the WRITE lock request blocked inside LockServer.lock(),
+         * this call should block until the first of these happens:
+         * -The READ lock times out and the WRITE lock can be granted, thus freeing up the debugLock
+         * -The logCurrentState tryLock() call times out after LOG_STATE_DEBUG_LOCK_WAIT_TIME_IN_MILLIS
+         *   and the call moves on to logCurrentStateInconsistent()
+         */
+        barrier.await();
```
[the remark this answers is not in the document]
Comment (dxiao) [reply lock-impl/src/test/java/com/palantir/lock/LockServiceTest.java:564]: Hmm. On reread, it's not needed at all. removing

**Comment by dxiao** [comment]
Updated! Let me know if there's anything else to do. Otherwise, would live to get this in time for the weekly release :)

**Review by jeremyk-91 — APPROVED** [review]
👍  let's merge this!

**Pull request merged**